File: Baekjoon/3665.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int n, m;
vector<vector<bool>> adj;
vector<int> ranking;
vector<int> in_degree;

void topological_sort() {
	vector<int> result;

	queue<int> q;
	for (int i = 1; i <= n; i++) {
		if (in_degree[i] == 0)
			q.push(i);
	}

	bool is_cycle = false;
	bool is_certain = true;

	for (int i = 0; i < n; i++) {
		if (q.empty()) {
			is_cycle = true;
			break;
		}
		else if (q.size() >= 2) {
			is_certain = false;
			break;
		}
		int cur = q.front();
		q.pop();
		result.push_back(cur);

		for (int j = 1; j <= n; j++) {
			if (adj[cur][j]) {
				in_degree[j]--;
				if (in_degree[j] == 0)
					q.push(j);
			}
		}
	}
	if (is_cycle)
		cout << "IMPOSSIBLE\n";
	else if (!is_certain)
		cout << "?\n";
	else {
		for (int i = 0; i < n; i++)
			cout << result[i] << ' ';
		cout << '\n';
	}
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0);

	int t;
	cin >> t;
	while (t--) {
		cin >> n;
		ranking = vector<int>(n + 1);
		in_degree = vector<int>(n + 1, 0);
		adj = vector<vector<bool>>(n + 1, vector<bool>(n + 1, false));
		for (int i = 1; i <= n; i++)
			cin >> ranking[i];

		for (int i = 1; i < n; i++)
			for (int j = i + 1; j <= n; j++) {
				adj[ranking[i]][ranking[j]] = true;
				in_degree[ranking[j]]++;
			}

		cin >> m;
		for (int i = 0; i < m; i++) {
			int a, b;
			cin >> a >> b;
			if (adj[a][b]) {
				adj[a][b] = false;
				adj[b][a] = true;
				in_degree[a]++;
				in_degree[b]--;
			}
			else {
				adj[a][b] = true;
				adj[b][a] = false;
				in_degree[a]--;
				in_degree[b]++;
			}
		}
		topological_sort();
	}
	return 0;
}

File: Baekjoon/3665.py
import sys
from collections import deque


def topological_sort():
    result = []
    q = deque()

    for i in range(1, n + 1):
        if in_degree[i] == 0:
            q.append(i)

    is_cycle = False
    is_certain = True

    for i in range(n):
        if len(q) == 0:
            is_cycle = True
            break
        elif len(q) >= 2:
            is_certain = False
            break
        cur = q.popleft()
        result.append(cur)

        for j in range(1, n + 1):
            if adj[cur][j]:
                in_degree[j] -= 1
                if in_degree[j] == 0:
                    q.append(j)
    if is_cycle:
        print("IMPOSSIBLE")
    elif not is_certain:
        print("?")
    else:
        for i in result:
            print(i, end=' ')
        print()


t = int(sys.stdin.readline().rstrip())
for _ in range(t):
    n = int(sys.stdin.readline().rstrip())
    in_degree = [0 for _ in range(n + 1)]
    adj = [[False] * (n + 1) for _ in range(n + 1)]
    ranking = list(map(int, sys.stdin.readline().rstrip().split()))

    for i in range(n - 1):
        for j in range(i + 1, n):
            adj[ranking[i]][ranking[j]] = True
            in_degree[ranking[j]] += 1
    m = int(sys.stdin.readline().rstrip())
    for _ in range(m):
        a, b = map(int, sys.stdin.readline().rstrip().split())
        if adj[a][b]:
            adj[a][b] = False
            adj[b][a] = True
            in_degree[a] += 1
            in_degree[b] -= 1
        else:
            adj[a][b] = True
            adj[b][a] = False
            in_degree[a] -= 1
            in_degree[b] += 1
    topological_sort()
